### PR TITLE
Move CLR_ADDITIONAL_COMPILER_OPTIONS to be set after default options

### PR DESCRIPTION
**Pull request by WardenGnaw**
This PR adds '-defaultvisibility' to build.sh.

This will define DEFAULT_VISIBILITY for CMake.
If DEFAULT_VISIBILITY is defined, it will not set the compile option to
have all methods default to be hidden.

### CODE REVIEW

**Comment by jkotas** [comment]
Could you please explain what this option is for?

**Comment by WardenGnaw** [comment]
@janvorli 

The debugger builds CoreCLR PAL and consume the resulting static libs into a package we produce. We use an export script to control the visibility of the PAL symbols used in the .so. 

This toggle allows us to build CoreCLR without having the default visibility being set hidden.

**Comment by janvorli** [comment]
@jkotas we have discussed various options with the debugger folks and it seems that this fits best their workflow so that they don't have to update pal.h any time they need a new export to be exposed.

File: configurecompiler.cmake [hunk 1]
```diff
@@ -492,8 +492,12 @@ if (CLR_CMAKE_PLATFORM_UNIX)
   # as x64 does. It has been causing issues in ARM (https://github.com/dotnet/coreclr/issues/4746)
   add_compile_options(-fsigned-char)
 
-  # We mark the function which needs exporting with DLLEXPORT
-  add_compile_options(-fvisibility=hidden)
+  # Check to see if build wants to keep visibility to default.
+  # Note: Once -defaultvisibility is set, it will be cached to always be true.
+  if(NOT DEFAULT_VISIBILITY)
```
Comment (janvorli) [inline configurecompiler.cmake]: Can you please rename the variable to CLR_CMAKE_DEFAULT_VISIBILITY?

**Comment by jkotas** [comment]
Can this just use `CLR_ADDITIONAL_COMPILER_OPTIONS` ?

**Comment by janvorli** [comment]
> Can this just use `CLR_ADDITIONAL_COMPILER_OPTIONS` ?

I prefer it the way it is done. If we used CLR_ADDITIONAL_COMPILER_OPTIONS for that, we would need to check its contents before adding `-fvisibility=hidden` option, which would be ugly.

**Comment by jkotas** [comment]
I do not think you need to check. The last option wins. No?

**Comment by janvorli** [comment]
> The last option wins. No?

It seems it does. However, the CLR_ADDITIONAL_COMPILER_OPTIONS is used before we set the visibility option, so we will at need to move it after we set all the explicit options. In fact, it would be good to move it anyways, since people may want to use it to override some of the options we set.

@WardenGnaw can you please update the change based on this? That means remove all the build.sh changes and modify the change in compileroptions.cmake by removing your change and moving the usage of the CLR_ADDITIONAL_COMPILER_OPTIONS from
https://github.com/dotnet/coreclr/blob/d9781f1d81b06ea1562521aea6d278ca876c82ad/configurecompiler.cmake#L326
to
https://github.com/dotnet/coreclr/blob/d9781f1d81b06ea1562521aea6d278ca876c82ad/configurecompiler.cmake#L518
(wrapped in the same `if` as in the original location)?

**Pull request merged**